app/Models/Extensions/PhotoCast.php

```diff
@@ -0,0 +1,204 @@
+		}
+
+		// Parse paths
+		$photo['thumbUrl'] = Storage::url('thumb/' . $this->thumbUrl);
+
+		if ($this->thumb2x == '1') {
+			$thumbUrl2x = Helpers::ex2x($this->thumbUrl);
+			$photo['thumb2x'] = Storage::url('thumb/' . $thumbUrl2x);
+		} else {
+			$photo['thumb2x'] = '';
+		}
+
+		$path_prefix = $this->type == 'raw' ? 'raw/' : 'big/';
+		$photo['url'] = Storage::url($path_prefix . $this->url);
+
+		if (isset($this->takestamp) && $this->takestamp != null) {
+			// Use takestamp
+			$photo['takedate'] = $this->takestamp->format('d F Y \a\t H:i');
+		} else {
+			$photo['takedate'] = '';
+		}
+
+		return $photo;
+	}
+
+	/**
+	 * Given a photo, return the proper URL.
+	 */
+	public function urls(array &$return): void
+	{
+		// if this is a video
+		if (strpos($this->type, 'video') === 0) {
+			$photoUrl = $this->thumbUrl;
+
+			// We need to format the framerate (stored as focal) -> max 2 decimal digits
+			$return['focal'] = round($return['focal'], 2);
+		} elseif ($this->type == 'raw') {
+			// It's a raw file -> we also use jpeg as extension
+			$photoUrl = $this->thumbUrl;
+		} else {
+			$photoUrl = $this->url;
+		}
+		$photoUrl2x = ($photoUrl !== '') ? Helpers::ex2x($photoUrl) : '';
+
+		$sizes = [
+			'medium' => $photoUrl,
+			'medium2x' => $photoUrl2x,
+			'small' => $photoUrl,
+			'small2x' => $photoUrl2x,
+		];
+
+		foreach ($sizes as $size => $url) {
+			if ($this->$size != '') {
+				$return[$size] = Storage::url(str_replace('2x', '', $size) . '/' . $url);
+				$return[$size . '_dim'] = $this->$size;
+			} else {
+				$return[$size] = '';
+				$return[$size . '_dim'] = '';
+			}
+		}
+
+		// Parse paths
+		$return['thumbUrl'] = Storage::url('thumb/' . $this->thumbUrl);
+		$return['thumb2x'] = ($this->thumb2x == '1') ? Storage::url('thumb/' . Helpers::ex2x($this->thumbUrl)) : '';
+
+		$path_prefix = ($this->type == 'raw') ? 'raw/' : 'big/';
+		$return['url'] = Storage::url($path_prefix . $this->url);
+
+		if ($this->livePhotoUrl !== '' && $this->livePhotoUrl !== null) {
+			$return['livePhotoUrl'] = Storage::url('big/' . $this->livePhotoUrl);
+		} else {
+			$return['livePhotoUrl'] = null;
+		}
+	}
+
+	/**
+	 * Given a Photo, returns the thumb version.
+	 */
+	public function toThumb(): Thumb
+	{
+		$symLinkFunctions = resolve(SymLinkFunctions::class);
+
+		$thumb = new Thumb($this->type, $this->id);
+		// maybe refactor?
+		$sym = $symLinkFunctions->find($this);
+		if ($sym !== null) {
+			$thumb->thumb = $sym->get('thumbUrl');
+			// default is '' so if thumb2x does not exist we just reply '' which is the behaviour we want
+			$thumb->thumb2x = $sym->get('thumb2x');
+		} else {
+			$thumb->thumb = Storage::url('thumb/' . $this->thumbUrl);
+			if ($this->thumb2x == '1') {
+				$thumb->set_thumb2x();
+			}
+		}
+
+		return $thumb;
+	}
+
+	/**
+	 * Downgrade the quality of the pictures.
+	 *
+	 * @param array $return
+	 */
+	public function downgrade(array &$return)
+	{
+		if (strpos($this->type, 'video') === 0) {
```

Comment:
Not sure. Because in the case of video the thumbnails are containing the preview.

Anyway, the check `strpos($this->type, 'video') === 0` should be integrated in the photo model (`$this->is_video()`)  making it more readable and easier to work with as we would make it a single application point to modify in case of further refactoring...

Reply:
Yeah, I saw multiple of these calls sprinkled under `app/Models`; I agree they should be replaced.

Back to the main point, I'm pretty sure I'm right. The job of `downgrade` is to prevent the full photo (`big`) from being available. It's not supposed to do anything for videos, because the only video variants we have are `big` (we don't resize video files, and we only generate `thumb` and `small` `.jpeg` previews for them).

I checked the code on `master` and there's a subtle difference: there it uses `== 0`. `strpos` returns `false` if there's no match, and since in PHP `false == 0`, the condition triggers for image files like it's supposed to. As far as I can tell though, that's just two bugs canceling each other out, rather than a deliberate coding :smiley:.